### ISSUE TITLE
[D8][SR] Add support for "trusted_host_patterns" (but do not enforce).

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2568]
https://www.drupal.org/node/1992030

> ## Trusted host security setting in Drupal 8
> 
> As of January 2015, [Drupal 8 supports "trusted host patterns"](https://www.drupal.org/node/2410395), where you can (and should) specify a set of regular expressions that the domains on incoming requests *must* match. Example configuration in settings.php would read:
> 
> ```php
> $settings['trusted_host_patterns'] = array(
>   '^www\.example\.com$',
> );
> ```
> 
> See the above change record for more details. Note that, if you're doing local development, you might get (temporarily) locked out of your site by the above configuration on its own. [You should add a trusted host pattern](http://drupal.stackexchange.com/a/145994/) for `'^localhost$'` in this case.

https://www.drupal.org/node/2410395

> ...
> **Description**: 
> It is possible to spoof the HTTP Host header for nefarious purposes, and trick Drupal into using a different domain name in several subsystems (particularly link generation). The Symfony framework provides a trusted host mechanism, where site administrators can whitelist hostnames. This mechanism can now be configured through settings.php.
>...

![d8-status_report-trusted_host_patterns](https://cloud.githubusercontent.com/assets/2423362/23728003/95a7f3f6-04ae-11e7-99e2-3846df0f8856.png)

---
Advocates: @klonos 

---

~~PR by @klonos: https://github.com/backdrop/backdrop/pull/2382~~
updated PR by @quicksketch: https://github.com/backdrop/backdrop/pull/2835

---

NOTE: The current PR relies on things being defined in settings.php, which we cannot edit in the PR sandboxes. So this will need to be tested locally:

1. Pull that PPR changes locally.
2. Run the Backdrop installer -> confirm no errors that would block the installation
3. Once the installation is finished, go to `/admin/reports/status` -> there should be a warning about trusted host patterns not being configured 👍 
4. Edit your `settings.php` file, and un-comment the `$settings['trusted_host_patterns'] = array('^www\.example\.com$');` line. Save.
5. Refresh the "Status report" page -> the warning should now be telling you that the host is not trusted according to the `trusted_host_patterns` setting, and also show you a list of currently configured patterns 👍 
6. Keep changing the `trusted_host_patterns` setting in your `settings.php` file, saving, and then refreshing the "Status report" page. Test what happens to the "Trusted Host Settings" entry with none/one/multiple patterns, and with a variety of matching and non-matching ones. Report back if it mis-reports trusted or untrusted hosts during your tests.

**Comment by klonos** [comment]
This is part of #378. I think that we should be providing a way to configure these in the UI and have a corresponding config file instead of (or in addition to) the settings.php file. In general I think that we should introduce a separate "Security" section under configuration and have any security related settings there. This could also include the settings of any future inclusion of modules like honeypot in core for example.

**Comment by klonos** [comment]
It turns out that this is being worked on for D7 too (but stale since Aug 2017): https://www.drupal.org/project/drupal/issues/2221699#comment-9599237

I have filed a PR for this. I have followed the same pattern as in Drupal and have set the message to be `'severity' => REQUIREMENT_ERROR`, but this makes it so that every Backdrop installation will have a red dot indicating a problem:

<img width="1016" alt="screen shot 2018-11-14 at 7 50 07 am" src="https://user-images.githubusercontent.com/2423362/48443462-54246b00-e7e5-11e8-85b3-7046f86449d5.png">

Is the security implication so severe, or should we make it a `REQUIREMENT_WARNING` instead?

**Comment by olafgrabienski** [comment]
@klonos It would be great if the trusted host could be set via the UI, even more if it's a red dot error.

**Comment by klonos** [comment]
As expected, it seems that I am not the first person to be concerned with certain aspects of this feature. For D8, the matter about enabling this setting during the installation is being tackled here: https://www.drupal.org/project/drupal/issues/2404259

Some comments of interest, related to the installer and UX:

> Ideally you shouldn't need to google a problem as soon as you've installed new software.
> 
> Yes there is a help message: "The trusted_host_patterns setting is not configured in settings.php." but at first contact (with Drupal) the user won't know where this file is or what it's for. A red error message will make them feel like something is very wrong and will likely assume their installation process, hosting config or something external to the software is possibly creating a conflict or is at fault.
> 
> If trusted_hosts being unset isn't a problem or should be optional then surely it makes sense to remove the warning/error and lower the warning level on the status page notification to prevent users thinking their installation is broken or they're doing something wrong.

>We also need to bear in my mind that a fair few hosters don't allow settings.php to be written during an install. See parts of [#2156401: Write install_profile value to configuration and only to settings.php if it is writeable](https://www.drupal.org/project/drupal/issues/2156401) for a long and big discussion of this.

>I think it is important that a user sees what will be registered as trusted host pattern.
I don't think the installer should do this silently.
>
> **Easy solution**
> Provide a checkbox in the installer: "Only allow requests from detected.hostname.com. This setting can later be changed in the settings.php under $settings['trusted_host_patterns']."
> The checkbox could be enabled by default.
> 
> **Advanced solution**
> Let the installer show a form element, where a user can define which hosts should be trusted.
> This should be pre-filled with the host from the current request.
> It could be a textarea, where each line is a host.
> Initially, this will not use regular expressions, but the plain host as-is. So it would say "example.com", not "^example\.com$".
> 
> There could be a checkbox or something to switch this to regular expression mode.
> But maybe this is not even needed. People who need those could do it in settings.php directly.
> 
> Add a description / help text, explaining that this will be converted into a regular expression in the settings.php, allowing to add additional trusted hosts later.

**Comment by klonos** [comment]
...so a few points to make here:

- We should consider demoting this to a warning instead of an error.
- We could be logging an error in the log each time there is a spoofing attempt (instead of completely blocking access to the site like in D8).
- We should not be blocking the installation if this setting is not enabled.

All the above would address the UX issues, because:

- the first-time user would not be greeted with red errors.
- there would not be an imperative need to provide a UI for configuring this setting, as it would become opt-in

Having said that, I am not sure of the severity of the security implications of not setting this. If we demote its importance and sites end up being hacked or identities and personal data stolen because of it, then we should definitely not have it be opt-in.

@serundeputy ? @quicksketch ?

**Comment by jenlampton** [comment]
Didn't we decide to do all the security things in 1.13? Changing milestone.

**Comment by klonos** [comment]
I would like to get this in 1.13.0, so i need some direction.

- Should we demote this to a warning? This might be better OOTB UX (no red errors), but could also be an excuse for others to say that we are not taking security seriously.

- Should we allow this to also be configured via the UI (#3624)? This can be done later btw.

**Comment by jenlampton** [comment]
> Should we demote this to a warning?

No, I think it should stay an error, we need the red dot.

**Comment by klonos** [comment]
That is great then, because I have refreshed the PR sandbox, and tests are all green still 😄

**Comment by olafgrabienski** [comment]
>> Should we demote this to a warning?

> No, I think it should stay an error, we need the red dot.

If it's an error including red dot, we need an easy way, ideally via the UI, or *really* good instructions to actually set the setting. Otherwise, we should make a warning of it.

(At the moment, in the message we link to the [drupal documentation](https://www.drupal.org/node/1992030) btw, which speaks of "Trusted host security setting in Drupal 8".)

**Comment by klonos** [comment]
> (At the moment, in the message we link to the drupal documentation btw, which speaks of "Trusted host security setting in Drupal 8".)

I have created a draft page in b.org here: https://backdropcms.org/trusted-host-settings. Please review so we can publish it once this gets in. I have assumed the following:

- This will get in 1.13
- The specific solutions for MAMP / XAMPP / Acquia Dev Desktop work (have not tested them myself)

I have updated the PR. In the message we show in the admin UI, I am pointing to b.org. In the docblock in `settings.php`, I have left the d.org link in place (in case we need to refer to it in the future), and have added the b.org link.

> If it's an error including red dot, we need an easy way, ideally via the UI, or really good instructions to actually set the setting. Otherwise, we should make a warning of it.

I'd really like this to get in 1.13. Adding a UI for this would be quite some work and can be done later, in a follow-up issue.

**Comment by olafgrabienski** [comment]
Thanks @klonos for updating the PR! I've had a look at https://backdropcms.org/trusted-host-settings. Ideas:

- Rename the section "Trusted host security setting in Backdrop 1.13.0" to "Trusted host security setting in Backdrop", i.e. without version number.
- Add the version number to the following sentence instead, e.g.: "As of version 1.13.0, Backdrop core supports "trusted host patterns" ..."
- Start the page with "Trusted host security setting in Backdrop". (As an alternative, point the link directly to the Trusted host section, but from the title of the page and from the link source I'd expect that the page starts with it.)
- Add the easy code example from `settings.php` (only `www.example.com`) to the page.
- Add another easy example to the page: only `example.com` (without `www`)
- Add the latter also to `settings.php`.

Note re formatting of the page: The code examples contain unwanted `&nbsp;` and `<br />` elements.

**Comment by klonos** [comment]
- [x] Rename the section "Trusted host security setting in Backdrop 1.13.0" to "Trusted host security setting in Backdrop", i.e. without version number.
- [x] Add the version number to the following sentence instead, e.g.: "As of version 1.13.0, Backdrop core supports "trusted host patterns" ..."
- [ ] Start the page with "Trusted host security setting in Backdrop". (As an alternative, point the link directly to the Trusted host section, but from the title of the page and from the link source I'd expect that the page starts with it.)
- [ ] Add the easy code example from `settings.php` (only `www.example.com`) to the page.
- [ ] Add another easy example to the page: only `example.com` (without `www`)
- [ ] Add the latter also to `settings.php`.

- [x] Note re formatting of the page: The code examples contain unwanted `&nbsp;` and `<br />` elements.

**Comment by klonos** [comment]
>Start the page with "Trusted host security setting in Backdrop".

Sorry @olafgrabienski, I don't understand what you are suggesting here 🤔 

I have fixed the "easy" things. Will need another pass when I find some time (unless anyone else beats me to it). We should get some more eyes on this.

**Comment by olafgrabienski** [comment]
> I don't understand what you are suggesting here

Ok, I'll try to explain ;-)

The page "Trusted Host settings" has the sections ordered as follows:

- (Intro)
- Scenario 1: ...
- Scenario 2: ...
- Scenario 3: ...
- Solutions to confusing experiences ...
- Trusted host security setting in Backdrop

So, we have a lot of information, but the most important one, the "Trusted host in Backdrop" section, comes at the very end of the page. The important information should stay at the top instead, for instance:

- (Intro)
- Trusted host security setting in Backdrop
- Other solutions
- Background information:
  - Scenario 1: ...
  - Scenario 2: ...
  - Scenario 3: ...

**Comment by jenlampton** [comment]
> If it's an error including red dot, we need an easy way, ideally via the UI, or really good instructions to actually set the setting. Otherwise, we should make a warning of it.

Oh, I see. Okay, you've convinced me :) Warning is fine.

**Comment by klonos** [comment]
> If it's an error including red dot, we need an easy way, ideally via the UI, or really good instructions to actually set the setting. Otherwise, we should make a warning of it.
>
>> Oh, I see. Okay, you've convinced me :) Warning is fine.

Yep, I think that having as a warning is fine for now. Once we have a UI (#3630) + well-documented instructions for how to set it up, we can then upgrade the warning to an error. I have updated the PR accordingly, and have added a `@todo`.

I will also update the docblock in settings.php to include more examples of patterns for common use cases.

**Comment by jenlampton** [comment]
Does this issue still need more feedback? 
Removing label (feel free to add back if it does!)

**Comment by olafgrabienski** [comment]
> Does this issue still need more feedback?

I don't think so :-) Apart of that:

- The documentation page on b.org has to be updated -- see https://github.com/backdrop/backdrop-issues/issues/2568#issuecomment-477604150. I've added the `needs - documentation` label.
- The PR needs testing. I've had a look at the sandbox site which looks fine. It's in a state where the *Trusted Host Settings* are not defined. I've however not tested the situation when the settings are defined (correctly or not).

**Comment by klonos** [comment]
I have tweaked the in-code comments/documentation a tiny bit.

> - The documentation page on b.org has to be updated...

This can be done post-commit. We have time till May 15 (and also after that) to tweak things.

In the meantime, I have updated the page in b.org to add a fragment:

to https://backdropcms.org/protecting-against-http-host-header-attacks#trusted_hosts

...I have also changed the link in settings php accordingly, and have removed the link to https://www.drupal.org/node/1992030.

PS: I think that our documentation pages should have a) fields to allow external references, and b) a "last updated" field/indication. That way, we can refer to the original documentation, and know when it might be time to revise

> - I've however not tested the situation when the settings are defined (correctly or not).

The current PR relies on things being defined in settings.php, which we cannot edit in the PR sandboxes. So this will need to be tested locally for now.

**Comment by jenlampton** [comment]
@klonos, I created an issue for b.org based on your suggestions: https://github.com/backdrop-ops/backdropcms.org/issues/533

**Comment by klonos** [comment]
> ...this is getting bumped.

Oh well, we tried 🙂

**Comment by jenlampton** [comment]
@klonos do you want to be the advocate for this issue for 1.14? Removing the milestone for now, but feel free to add it back if so!

**Comment by klonos** [comment]
Yep, I'll start with this one. Hoping for this to get in relatively soon, so I can then advocate another issue 😄

**Comment by jenlampton** [comment]
@klonos can you please update the link in the status report to go to https://api.backdropcms.org/documentation/protecting-against-http-host-header-attacks?

**Comment by klonos** [comment]
Done! 😉

**Comment by quicksketch** [comment]
I left some feedback in https://github.com/backdrop/backdrop/pull/2382#pullrequestreview-232809885.

**Comment by klonos** [comment]
@quicksketch I've updated the PR.

**Comment by klonos** [comment]
@stpaultim and anyone else interested to test this, I have added steps in the issue summary 😉

**Comment by klonos** [comment]
OK, I have finished all I could think of for this one: https://github.com/backdrop/backdrop/pull/2382

Screenshots...

No `trusted_host_patterns` configured:

<img width="1189" alt="Screen Shot 2019-09-01 at 6 51 38 am" src="https://user-images.githubusercontent.com/2423362/64071558-5cbb3480-cc85-11e9-953a-edeeb15f2f5b.png">

A single, non-matching pattern:

<img width="1182" alt="Screen Shot 2019-09-01 at 6 57 25 am" src="https://user-images.githubusercontent.com/2423362/64071587-12868300-cc86-11e9-8d15-d2abe4d22305.png">

Multiple non-matching patterns:

<img width="1188" alt="Screen Shot 2019-09-01 at 6 58 43 am" src="https://user-images.githubusercontent.com/2423362/64071588-17e3cd80-cc86-11e9-937c-e0c0c57b09cd.png">

A single matching pattern:

<img width="1181" alt="Screen Shot 2019-09-01 at 6 59 23 am" src="https://user-images.githubusercontent.com/2423362/64071591-1ca88180-cc86-11e9-9847-e29302d7815d.png">

Mix of multiple, matching and non-matching patterns:

<img width="1196" alt="Screen Shot 2019-09-01 at 6 50 18 am" src="https://user-images.githubusercontent.com/2423362/64071568-93914a80-cc85-11e9-8f5f-391a759c42c8.png">

**Comment by klonos** [comment]
...I have also updated the draft in https://api.backdropcms.org/documentation/trusted-host-settings earlier today to say "Since Backdrop 1.14.0" instead of "1.13.0" 😉

**Comment by quicksketch** [comment]
Hey @klonos, I made minor modifications and filed a new PR at https://github.com/backdrop/backdrop/pull/2835 that makes a few changes:

- Rewords the status message to make it less aggressive.
- Fixed the `@since` line (there was a stray asterisk in there).
- I added the ability to set `$settings['trusted_host_patterns'] = FALSE;` which will downgrade the warning to an "info" message.

The (unpublished) documentation at https://api.backdropcms.org/documentation/trusted-host-settings needs some work. I think for most people the only thing they'll want is to figure out how to fix the warning: the "how" and not the "why". But we can work on that any time of course, it's not time-sensitive.

Lastly, I _just_ discovered that this PR doesn't yet provide any actual protection! If you put an invalid host name in settings.php, like `www\.example\.com`, traffic is *not blocked*. Seems like the message and everything works, but not the protection functionality. This is because `backdrop_environment_initialize()` (where the hostname is checked) happens before `backdrop_settings_initialize()` (see `_backdrop_bootstrap_configuration() where both things happen). We'll either have to rearrange the bootstrap slightly or add a second check after settings.php is read in.

**Comment by quicksketch** [comment]
So to follow-up a little further, in D8 when this setting is enabled, it blocks _all_ traffic to the site. Nothing is accessible (even the log-in) if you try to access a site when the hostname is incorrectly configured. This _would_ be the case in our current patch if the execution order were fixed. There is _zero_ messaging currently in Backdrop. Here's the section of code that would execute:

```
  if (isset($_SERVER['HTTP_HOST'])) {
    // As HTTP_HOST is user input, ensure it only contains characters allowed
    // in hostnames. See RFC 952 (and RFC 2181).
    // $_SERVER['HTTP_HOST'] is lowercased here per specifications.
    $_SERVER['HTTP_HOST'] = strtolower($_SERVER['HTTP_HOST']);
    if (!backdrop_valid_http_host($_SERVER['HTTP_HOST'])) {
      // HTTP_HOST is invalid, e.g. if containing slashes it may be an attack.
      header($_SERVER['SERVER_PROTOCOL'] . ' 400 Bad Request');
      exit;
    }
  }
```

So you'd get an HTTP 400 error and a blank white page. Obviously that needs to be improved to include some kind of explanation, preferably better than Drupal's which gives you a very ambiguous error, `"The provided host name is not valid for this server."` which isn't at all obvious that it's a Drupal configuration problem and not something wrong with your web host.

**Comment by quicksketch** [comment]
Note if we did the same thing as D8 (blocking all traffic) then messaging for when the hostname does not match isn't necessary in `system_requirements()`, because a user isn't going to possibly see that page if they can't access the entire site.

**Comment by quicksketch** [comment]
I updated my PR at https://github.com/backdrop/backdrop/pull/2835 so that it now blocks traffic (with a message) if the provided HTTP Host is not accepted:

![image](https://user-images.githubusercontent.com/100206/64081691-36af8780-ccb9-11e9-829e-c7ae600ca4ab.png)

**Comment by klonos** [comment]
I have tested this on my local, and it works as expected 👍 ...I have left 2 code suggestions in the PR re wording, but not a biggie.

PS: I have tested running the installer with an intentionally misconfigured pattern in `settings.php`. I got redirected to `/core/install.php`, where there was a default `ERR_TOO_MANY_REDIRECTS` browser error. Not sure if we still want to show a friendly message when `$phase == 'install'` too, but can be a follow-up if we decide to.

**Comment by quicksketch** [comment]
>  I have tested running the installer with an intentionally misconfigured pattern in settings.php. I got redirected to /core/install.php, where there was a default ERR_TOO_MANY_REDIRECTS browser error.

Updated the PR to work-around this issue. Similar to problems with the config directory, we suppress errors if the user is in the installer. This means that the install will complete perfectly fine, but then once you are taken to your site you'll get the error. Not friendly, but at least the situation is explained with a message instead of an infinite redirect loop.

I also reworded the settings.php documentation further and moved `$settings['trusted_host_patterns']` a lot higher in the file, since this setting is now one of the most likely things to need changing manually.

**Comment by quicksketch** [comment]
I merged in https://github.com/backdrop/backdrop/pull/2835 into 1.x for 1.14.0. Thanks @klonos for all your work on this! Let's file follow-ups if needed.

**Comment by olafgrabienski** [comment]
Reminder: the documentation page https://api.backdropcms.org/documentation/trusted-host-settings is still unpublished. What's the best time to publish it?

**Comment by klonos** [comment]
@olafgrabienski I think closer to the 15th, just before the release. I have added it to the tasks in #4005 so we don't forget 😉

**Comment by domaingood** [comment]
Not configured less
<?php
/**
 * @file
 * Configuration file for the Backdrop CMS multi-site feature.
 *
 * This file allows you to map hostnames, ports, and pathnames to configuration
 * directories. Information on how to set up multisites see the README.md file
 * in this directory.
 */
// $sites['example.com'] = 'example';
$sites['phalconcms.com'] = 'phalconcms';

phalconcms.com/sites/sites.php

Did I missing something

**Comment by klonos** [comment]
@domaingood can you please explain what the issue is? Is that not working for you? I would either open a separate issue, of create a post in https://forum.backdropcms.org and add more details there.
